src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -389,7 +398,8 @@ private void signalIfAllRegisteredInQueue() {
         // is active at a time - a caller of lock() that has just acquired
         // the pool lock but is waiting for all registered elements to be
         // returned to the queue.
-        if (registeredElements.size() == liveQueue.size()) {
+        if (registeredElements.size() == liveQueue.size() ||
+                poisonPill != null) {
```

Comment:
That work is probably still pending, but I had been thinking that in the `lock()` method we would be throwing an exception from the `lock()` method -- and, if necessary, freeing the pool lock -- if a poison pill exists.  Inside of the current `lock()` code, we can get stuck in an `await` if the pool is not full:

```clj
while (registeredElements.size() != liveQueue.size()) {
  allRegisteredInQueue.await();
}
```
With the new pill logic, I think we need to have that condition variable signaled when a pill is inserted into the pool so that the `lock()` caller doesn't get stuck indefinitely.  I had presumed that's what the extra check above was for.  Is that not right?

For clarity, I think it might be good to rename `allRegisteredInQueue` to something which indicates that a pill can disrupt a caller who is waiting for that to be signalled.  Maybe something like `pillInsertedOrAllRegisteredInQueue`.  Ugh, horrible naming... :(